tools/taish/tai_shell.cpp

```diff
@@ -0,0 +1,593 @@
+  m_ostr = new std::ostream(m_ofilebuf);
+  return m_client_fd;
+}
+
+void tai_cli_server::disconnect() {
+  delete m_istr;
+  delete m_ostr;
+  delete m_ifilebuf;
+  delete m_ofilebuf;
+  close (m_client_fd);
+  m_client_fd = -1;
+}
+
+int tai_cli_server::recv() {
+  return cmd_parse (m_istr, m_ostr);
+}
+
+int tai_cli_shell::cmd_parse(std::istream *istr, std::ostream *ostr) {
+  int ret = 0;
+  std::string buf;
+  std::vector <std::string> *args;
+  std::map<std::string, tai_command_fn>::iterator cmd;
+
+  std::getline (*istr, buf);
+  if (istr->eof() || istr->fail())
+    return -10;
+
+  if (buf.back() == '\r') { // remove CR
+    buf.resize(buf.size()-1);
+  }
+
+  args = make_args (buf);
+  if (args != nullptr && args->size() != 0) {
+    cmd = cmd2handler.find((*args)[0]);
+    if (cmd != cmd2handler.end()) {
+      ret = cmd->second(ostr, args);
+    } else {
+      *ostr << "unknown command(" << (*args)[0] << ") was speified!!\n";
+    }
+  }
+
+  if (args != nullptr) {
+    delete args;
+  }
+
+  *ostr << "> ";
+  ostr->flush();
+  return ret;
+}
+
+std::vector <std::string> *tai_cli_shell::make_args (std::string cmd) {
+  int length;
+  char *arg;
+  std::vector <std::string> *args;
+  char *saveptr = nullptr;
+
+  args = new std::vector <std::string>;
+  if (args == nullptr) {
+    return nullptr;
+  }
+  arg = strtok_r ((char *)cmd.c_str(), " ", &saveptr);
+  while (arg != nullptr) {
+    args->push_back (std::string(arg, strlen(arg)));
+    arg = strtok_r (NULL, " ", &saveptr);
+  }
+
+  return args;
+}
+
+tai_api::tai_api (void *lib_handle) {
+  initialize        = (tai_status_t (*)( _In_ uint64_t, _In_ const tai_service_method_table_t *))
+                         dlsym(lib_handle, "tai_api_initialize");
+  uninitialize      = (tai_status_t (*)(void))
+                         dlsym(lib_handle, "tai_api_uninitialize");
+  query             = (tai_status_t (*)( _In_ tai_api_t, _Out_ void**))
+                         dlsym(lib_handle, "tai_api_query");
+  log_set           = (tai_status_t (*)( _In_ tai_api_t, _In_ tai_log_level_t))
+                         dlsym(lib_handle, "tai_log_set");
+  object_type_query = (tai_object_type_t (*)( _In_ tai_object_id_t))
+                         dlsym(lib_handle, "tai_object_type_query");
+  module_id_query   = (tai_object_id_t (*)( _In_ tai_object_id_t))
+                         dlsym(lib_handle, "tai_module_id_query");
+  dbg_generate_dump = (tai_status_t (*)(_In_ const char *))
+                         dlsym(lib_handle, "tai_dbg_generate_dump");
+  log_level = TAI_LOG_LEVEL_INFO;
+}
+
+int tai_command_load (std::ostream *ostr, std::vector <std::string> *args) {
+  void *lib_handle;
+
+  if (args->size() != 2) {
+    *ostr << "Invalid parameters were specified!!\n";
+    return -1;
+  }
+
+  if (p_tai_api != nullptr) {
+    *ostr << "TAI library is already loaded!!\n";
+    return -1;
+  }
+
+  lib_handle = dlopen ((*args)[1].c_str(), RTLD_NOW | RTLD_GLOBAL);
```

Comment:
I don't think we want to encourage applications binding to the individual adapters in a modular system. All applications should bind via the libtai.so. In a modular system that would be the tai_mux.

Reply:
@scott-emery I don't have a strong preference here.
I think the main use of this app is a debug/develop tool for developers who are trying to use `libtai.so` provided by vendors.
In that case, it could be useful to load/unload `libtai.so` dynamically. WDYT?

Reply:
We should make it clear, then, that this is a development tool. I just didn't want other TAI host implementations to use this as an example and replicate this method of loading the TAI adapter library. Additionally, the tool makes some assumptions about which attributes need to be set and the values for those attributes (for example, in module::create_netif() the grid spacing and modulation are set to values which may not be supported on all modules). If we make it clear that this is a development tool, which will likely have to be modified depending on the target environment, then I'm OK with including it in this repo and using dlopen() to link the TAI adapter module.

Reply:
@scott-emery I agree with you. 
Since `TAI_NETWORK_INTERFACE_ATTR_INDEX`  is the only mandatory attribute for netif for now, I think it is better not to set other attributes here.

We can add a command later to set/get arbitrary attributes to TAI objects.